src/contracts/exchange/NiftyswapExchange20.sol

```diff
@@ -0,0 +1,810 @@
+  address internal factory;                       // address for the factory that created this contract
+  uint256 internal constant FEE_MULTIPLIER = 995; // Multiplier that calculates the fee (0.5%)
+
+  // Mapping variables
+  mapping(uint256 => uint256) internal totalSupplies;    // Liquidity pool token supply per Token id
+  mapping(uint256 => uint256) internal currencyReserves; // currency Token reserve per Token id
+
+  /***********************************|
+  |            Constructor           |
+  |__________________________________*/
+
+  /**
+   * @notice Create instance of exchange contract with respective token and currency token
+   * @param _tokenAddr     The address of the ERC-1155 Token
+   * @param _currencyAddr  The address of the ERC-20 currency Token
+   */
+  constructor(address _tokenAddr, address _currencyAddr) public {
+    require(
+      address(_tokenAddr) != address(0) && _currencyAddr != address(0),
+      "NiftyswapExchange20#constructor:INVALID_INPUT"
+    );
+    factory = msg.sender;
+    token = IERC1155(_tokenAddr);
+    currency = _currencyAddr;
+  }
+
+  /***********************************|
+  |        Exchange Functions         |
+  |__________________________________*/
+
+  /**
+   * @notice Convert currency tokens to Tokens _id and transfers Tokens to recipient.
+   * @dev User specifies MAXIMUM inputs (_maxCurrency) and EXACT outputs.
+   * @dev Assumes that all trades will be successful, or revert the whole tx
+   * @dev Exceeding currency tokens sent will be refunded to recipient
+   * @dev Sorting IDs is mandatory for efficient way of preventing duplicated IDs (which would lead to exploit)
+   * @param _tokenIds             Array of Tokens ID that are bought
+   * @param _tokensBoughtAmounts  Amount of Tokens id bought for each corresponding Token id in _tokenIds
+   * @param _maxCurrency          Total maximum amount of currency tokens to spend for all Token ids
+   * @param _deadline             Timestamp after which this transaction will be reverted
+   * @param _recipient            The address that receives output Tokens and refund
+   * @return currencySold How much currency was actually sold.
+   */
+  function _currencyToToken(
+    uint256[] memory _tokenIds,
+    uint256[] memory _tokensBoughtAmounts,
+    uint256 _maxCurrency,
+    uint256 _deadline,
+    address _recipient)
+    internal nonReentrant() returns (uint256[] memory currencySold)
+  {
+    // Input validation
+    require(_deadline >= block.timestamp, "NiftyswapExchange20#_currencyToToken: DEADLINE_EXCEEDED");
+
+    // Number of Token IDs to deposit
+    uint256 nTokens = _tokenIds.length;
+    uint256 totalRefundCurrency = _maxCurrency;
+
+    // Initialize variables
+    currencySold = new uint256[](nTokens); // Amount of currency tokens sold per ID
+    uint256[] memory tokenReserves = new uint256[](nTokens);  // Amount of tokens in reserve for each Token id
```

Comment:
I am fairly sure this array is never used, I think we can replace this line and the following one with just:

`uint256[] memory tokenReserves = _getTokenReserves(_tokenIds);`

Reply:
Sorry I had this in for debugging. Will remove